wallet/dapps-communication.md

```diff
@@ -8,4 +8,148 @@ Initial meta issue: <https://github.com/massalabs/massa-standards/issues/13>
 
 ## Specification
 
+Here is pseudo code for the two points of view:
+
+- browser extension developer
+- dapp builder
+
+**As a browser extension developer, I would write...**
+
+browser extension code, [content script](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts):
+
+```text
+// the extension register itself
+if (window.massaWalletProvider == undefined) { /* wait */ }
+
+window.bearbyWalletProvider = new EventTarget()
+
+window.massaWalletProvider.addEventListener('loaded', () => {
+  window.massaWalletProvider.dispatchEvent(new CustomEvent('register', {
+    ProviderName: "bearby",
+    eventTarget: "bearbyWalletProvider",
+    extensionID: "" // maybe not needed
+  }))
+})
+
+window.bearbyWalletProvider.addEventListener('sign', () => {
+  // validate the inputs
+  ...
+
+  // ask the background script to sign
+  browser.runtime.sendMessage({...})
+  .then((pkey, signature) => {
+    // send back the signature to the page script (massa js library)
+    window.bearbyWalletProvider.dispatchEvent(new CustomEvent('signed', {
+      pkey,
+      signature,
+      correlationId
+    }))
+  });
+})
+```
+
+browser extension code, [background script](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts):
+
+```text
+browser.runtime.onMessage.addListener((message, sender, response) => {
+  switch massage.command.name:
+    case 'sign':
+      return new Promise() // aks user authorization and password, sign message payload, return pkey and signature
+      break
+});
+```
+
+**Code snippet of massa javascript (or typescript) library:**
+
+```text
+registeredProviders = []
+actions = [] // all the ongoing actions (sign, getBalance...)
+
+// we start by listening register event
+window.massaWalletProvider = new EventTarget()
+window.massaWalletProvider.addEventListener('register', (payload) => {
+  // add the new provider to the list
+  // add the attribute wallets: list of wallets of this new provider
+  // the first wallet is a newly created wallet, we give as constructor argument the event target name created by the extension
+  registeredProviders.push({...payload, wallets: [new Wallet(payload.eventTarget)])
+})
+
+window.massaWalletProvider.dispatchEvent('loaded') // all wallet will catch this event and emit register event
+
+export class Wallet {
+  construct(private eventTargetName) {}
+
+  function sign(payload) {
+    const action = {
+      correlationId: uuid() // pick a correlation id to track the sign message request, it can simply be a int incremented
+      payload,
+      response: null
+    }
+    actions.push(action)
+
+    // listen for the response
+    window[eventTargetName].addEventListener('signed', ({ pkey, signature, correlationId }) => {
```

Comment:
the problem here is that we add a listener at each Wallet.sign, we must either add only one listener for the life of the wallet (for example in the constructor), either remove the listener when we have the response (for this we can use parameter `once`).